zkevm-circuits/src/evm_circuit/execution/error_oog_log.rs

```diff
@@ -0,0 +1,366 @@
+use crate::{
+    evm_circuit::{
+        execution::ExecutionGadget,
+        param::{N_BYTES_GAS, N_BYTES_MEMORY_WORD_SIZE},
+        step::ExecutionState,
+        util::{
+            common_gadget::RestoreContextGadget,
+            constraint_builder::{
+                ConstraintBuilder, StepStateTransition,
+                Transition::{Delta, Same},
+            },
+            math_gadget::LtGadget,
+            memory_gadget::{MemoryAddressGadget, MemoryExpansionGadget},
+            CachedRegion, Cell,
+        },
+        witness::{Block, Call, ExecStep, Transaction},
+    },
+    table::CallContextFieldTag,
+    util::Expr,
+};
+use eth_types::Field;
+use eth_types::{evm_types::GasCost, evm_types::OpcodeId};
+use halo2_proofs::{circuit::Value, plonk::Error};
+
+#[derive(Clone, Debug)]
+pub(crate) struct ErrorOOGLogGadget<F> {
+    opcode: Cell<F>,
+    // memory address
+    memory_address: MemoryAddressGadget<F>,
+    // constrain gas left is less than gas cost
+    memory_expansion: MemoryExpansionGadget<F, 1, N_BYTES_MEMORY_WORD_SIZE>,
+    insufficient_gas: LtGadget<F, N_BYTES_GAS>,
+    rw_counter_end_of_reversion: Cell<F>,
+    restore_context: RestoreContextGadget<F>,
+}
+
+impl<F: Field> ExecutionGadget<F> for ErrorOOGLogGadget<F> {
+    const NAME: &'static str = "ErrorOutOfGasLOG";
+
+    const EXECUTION_STATE: ExecutionState = ExecutionState::ErrorOutOfGasLOG;
+
+    fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
+        let opcode = cb.query_cell();
+        // constrain op code is Log*
+        cb.require_in_set(
+            "ErrorOutOfGasLOG happens in Log*",
+            opcode.expr(),
+            vec![
+                OpcodeId::LOG0.expr(),
+                OpcodeId::LOG1.expr(),
+                OpcodeId::LOG2.expr(),
+                OpcodeId::LOG3.expr(),
+                OpcodeId::LOG4.expr(),
+            ],
+        );
+        let mstart = cb.query_cell_phase2();
+        let msize = cb.query_word_rlc();
+        let rw_counter_end_of_reversion = cb.query_cell();
+
+        // Pop mstart_address, msize from stack
+        cb.stack_pop(mstart.expr());
+        cb.stack_pop(msize.expr());
+
+        // Note: no need to check not in static call, since write protection error will
+        // handle it.
+        let topic_count = opcode.expr() - OpcodeId::LOG0.as_u8().expr();
+
+        // check memory
+        let memory_address = MemoryAddressGadget::construct(cb, mstart, msize);
+
+        // Calculate the next memory size and the gas cost for this memory
+        // access
+        let memory_expansion = MemoryExpansionGadget::construct(cb, [memory_address.address()]);
+
+        let gas_cost = GasCost::LOG.as_u64().expr()
+            + GasCost::LOG.as_u64().expr() * topic_count
```

Comment:
This ends up in `cost(LOG_N) * (N+1)`. Right?? Why don't we do:
```suggestion
        let gas_cost = GasCost::LOG.as_u64().expr() * topic_count
```

Reply:
no, if opcode is LOG0 , `topic_count` is zero, then `gas_cost` will be zero by `let gas_cost = GasCost::LOG.as_u64().expr() * topic_count`

Reply:
Ohh that makes sense! Thanks!